File: ios/Classes/FlutterZoomWrapperPlugin.swift
import Flutter
import UIKit
import MobileRTC

public class FlutterZoomWrapperPlugin: NSObject, FlutterPlugin,MobileRTCAuthDelegate {
    var resultHolder: FlutterResult?


  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_zoom_wrapper", binaryMessenger: registrar.messenger())
    let instance = FlutterZoomWrapperPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

   public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
         switch call.method {
         case "getPlatformVersion":
             result("iOS " + UIDevice.current.systemVersion)
         case "initZoom":
             if let args = call.arguments as? [String: Any],
                let jwtToken = args["jwt"] as? String {
                 initializeZoom(jwt: jwtToken, result: result)
             } else {
                 result(FlutterError(code: "INVALID_ARGUMENT", message: "JWT token is missing", details: nil))
             }
         case "joinMeeting":
             if let args = call.arguments as? [String: Any],
                let meetingId = args["meetingId"] as? String,
                let password = args["meetingPassword"] as? String,
                let displayName = args["displayName"] as? String {
                 joinMeeting(meetingId: meetingId, password: password, displayName: displayName, result: result)
             } else {
                 result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing meeting parameters", details: nil))
             }
         default:
             result(FlutterMethodNotImplemented)
         }
     }

    private func initializeZoom(jwt: String, result: @escaping FlutterResult) {
        resultHolder = result

        // Initialize Zoom SDK
        let context = MobileRTCSDKInitContext()
        context.domain = "zoom.us"
        context.enableLog = true

        guard MobileRTC.shared().initialize(context) else {
            result(FlutterError(code: "INIT_FAILED", message: "Zoom SDK initialization failed", details: nil))
            return
        }

        guard let authService = MobileRTC.shared().getAuthService() else {
            result(FlutterError(code: "AUTH_SERVICE_NULL", message: "Zoom AuthService is nil", details: nil))
            return
        }

        authService.delegate = self
        authService.jwtToken = jwt
        authService.sdkAuth()
    }


         public func onMobileRTCAuthReturn(_ returnValue: MobileRTCAuthError) {
                 if returnValue == .success {
                     resultHolder?(true)
                 } else {
                     resultHolder?(FlutterError(code: "AUTH_FAILED", message: "Zoom SDK auth failed with code: \(returnValue.rawValue)", details: nil))
                 }
                 resultHolder = nil
             }

             private func joinMeeting(meetingId: String, password: String, displayName: String, result: @escaping FlutterResult) {
                     guard let meetingService = MobileRTC.shared().getMeetingService() else {
                         result(FlutterError(code: "NO_MEETING_SERVICE", message: "Meeting service is not available", details: nil))
                         return
                     }

                     let joinParams = MobileRTCMeetingJoinParam()
                     joinParams.userName = displayName
                     joinParams.meetingNumber = meetingId
                     joinParams.password = password

                     let response = meetingService.joinMeeting(with: joinParams)
                     result(response == MobileRTCMeetError.success)
                 }
}

File: android/src/main/kotlin/com/mfy/flutter_zoom_wrapper/FlutterZoomWrapperPlugin.kt
package com.mfy.flutter_zoom_wrapper

import android.content.Context
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import us.zoom.sdk.*
import us.zoom.sdk.ZoomError


class FlutterZoomWrapperPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, ZoomSDKInitializeListener {

  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private lateinit var zoomSDK: ZoomSDK
  private var activityBinding: ActivityPluginBinding? = null
  private var pendingStatus: Boolean? = false

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_zoom_wrapper")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
    zoomSDK = ZoomSDK.getInstance()
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
      "initZoom" -> {
        val jwt = call.argument<String>("jwt")
        if (jwt.isNullOrEmpty()) {
          result.error("INVALID_ARGUMENT", "JWT token is missing", null)
        } else {
          initZoom(jwt, result)
        }
      }
      "joinMeeting" -> {
        val meetingId = call.argument<String>("meetingId")
        val password = call.argument<String>("meetingPassword")
        val displayName = call.argument<String>("displayName")
        joinMeeting(meetingId, password, displayName, result)
      }
      else -> result.notImplemented()
    }
  }

  private fun initZoom(jwt: String, result: MethodChannel.Result) {
    //Log.d("ZOOM_SDK", "JWT Token received: $jwt")
   // Log.d("ZOOM_SDK", "JWT Token length: ${jwt.length}")
    zoomSDK = ZoomSDK.getInstance()

    if (zoomSDK.isInitialized) {
      result.success(true)
      return
    }

    val activity = activityBinding?.activity
    if (activity == null) {
      result.error("NO_ACTIVITY", "Activity is null. Cannot initialize Zoom SDK.", null)
      return
    }

    val initParams = ZoomSDKInitParams().apply {
      jwtToken = jwt
      domain = "zoom.us"
      enableLog = true
      enableGenerateDump = true
      logSize = 5
    }

    val listener = object : ZoomSDKInitializeListener {
      override fun onZoomSDKInitializeResult(errorCode: Int, internalErrorCode: Int) {
        if (errorCode == ZoomError.ZOOM_ERROR_SUCCESS) {
          result.success(true)
        } else {
          result.error("INIT_ERROR", "Failed to initialize Zoom SDK. Error: $errorCode, internalErrorCode: $internalErrorCode", null)
        }
      }

      override fun onZoomAuthIdentityExpired() {
        Log.w("Zoom", "Auth identity expired")
      }
    }

    zoomSDK.initialize(activity, listener, initParams)
  }

  private fun initZooms(jwt: String, result: MethodChannel.Result) {
    Log.d("ZOOM_SDK", "JWT Token received: $jwt")
    Log.d("ZOOM_SDK", "JWT Token length: ${jwt.length}")
    zoomSDK = ZoomSDK.getInstance()

    try {
      // Check if SDK is already initialized
      if (zoomSDK.isInitialized()){
        Log.d("ZOOM_SDK", "Zoom SDK already initialized")
        result.success(true)
      }
      // Only initialize if not already initialized
      if (!zoomSDK.isInitialized()) {
        // Configure SDK initialization parameters
        var initParams =  ZoomSDKInitParams()
        initParams.jwtToken = jwt;
        initParams.enableLog = true;
        initParams.enableGenerateDump = true;
        initParams.logSize = 5;
        initParams.domain="zoom.us";
        initParams.videoRawDataMemoryMode = ZoomSDKRawDataMemoryMode.ZoomSDKRawDataMemoryModeStack;

        // Create listener for initialization callbacks
        val listener = object : ZoomSDKInitializeListener {
          /**
           * Callback for SDK initialization result
           * @param errorCode Error code from initialization
           * @param internalErrorCode Internal error code from initialization
           */
          override fun onZoomSDKInitializeResult(errorCode: Int, internalErrorCode: Int) {
            if (errorCode != ZoomError.ZOOM_ERROR_SUCCESS) {
              Log.d("Failed", "Failed to initialize Zoom SDK")
              Log.d(
                "Failed to initialize Zoom SDK. Error: \" + $errorCode + \"",
                "internalErrorCode=\" + $internalErrorCode"
              )
              return result.error("INIT_ERROR", "\"Failed to initialize Zoom SDK. Error: \" + $errorCode + \",\n" +
                      "\"internalErrorCode=\" + $internalErrorCode\"", null)
            } else {
              Log.d("Success", "Initialize Zoom SDK successfully.")
              return result.success(true)
            }
          }

          /**
           * Callback when Zoom auth identity expires
           */
          override fun onZoomAuthIdentityExpired() {

          }
        }

        // Initialize SDK with prepared parameters
        zoomSDK.initialize(context, listener, initParams)

      }
    } catch (e: Exception) {
      Log.e("ZOOM_SDK", "Error while initializing ZoomSDK: ${e.message}", e)
      result.error("INIT_ERROR", "Failed to initialize Zoom SDK: ${e.message}", null)
    }
  }




  private fun joinMeeting(meetingId: String?, password: String?, displayName: String?, result: Result) {
    if (!zoomSDK.isInitialized) {
      result.error("SDK_NOT_INITIALIZED", "Zoom SDK not initialized", null)
      return
    }

    if (meetingId.isNullOrEmpty() || password.isNullOrEmpty() || displayName.isNullOrEmpty()) {
      result.error("INVALID_ARGUMENTS", "Missing meeting details", null)
      return
    }

    val joinParams = JoinMeetingParams().apply {
      meetingNo = meetingId
      this.password = password
      this.displayName = displayName
    }

    val options = JoinMeetingOptions()

    val meetingService = zoomSDK.meetingService
    activityBinding?.activity?.let {
      meetingService.joinMeetingWithParams(it, joinParams, options)
    } ?: run {
      meetingService.joinMeetingWithParams(context, joinParams, options)
    }

    result.success(true)
  }

  override fun onZoomSDKInitializeResult(errorCode: Int, internalErrorCode: Int) {
    when (errorCode) {
      ZoomError.ZOOM_ERROR_SUCCESS -> {
        Log.d("Zoom", "✅ Zoom SDK initialized successfully.")
      }
      1001 -> { // Numeric value for ZOOM_ERROR_WRONG_ZOOM_DOMAIN
        Log.e("Zoom", "❌ Wrong Zoom domain configured")
      }
      else -> {
        Log.e("Zoom", "❌ Zoom SDK initialization failed. Error code: $errorCode, Internal error: $internalErrorCode")
      }
    }
  }



  override fun onZoomAuthIdentityExpired() {
    Log.w("Zoom", "Auth identity expired")
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    Log.d("ZoomPlugin", "Attached to activity: ${binding.activity}")
    activityBinding = binding
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activityBinding = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activityBinding = binding
  }

  override fun onDetachedFromActivity() {
    activityBinding = null
  }
}
